jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c

```diff
@@ -1562,51 +1563,61 @@ ecma_builtin_array_prototype_object_unshift (const ecma_value_t args[], /**< arg
     }
   }
 
-  /* 4. */
-  if (len + args_number > ECMA_NUMBER_MAX_SAFE_INTEGER)
-  {
-    return ecma_raise_type_error (ECMA_ERR_MSG ("Unshift elements over 2**53-1 length is disallowed"));
-  }
-
-  /* 5. and 6. */
-  for (ecma_length_t k = len; k > 0; k--)
+#if ENABLED (JERRY_ESNEXT)
+  const bool should_iterate = args_number > 0;
+#else /* !ENABLED (JERRY_ESNEXT) */
+  const bool should_iterate = true;
+#endif /* ENABLED (JERRY_ESNEXT) */
+  if (should_iterate)
   {
-    /* 6.a, 6.c*/
-    ecma_value_t get_value = ecma_op_object_find_by_index (obj_p, k - 1);
-
-    if (ECMA_IS_VALUE_ERROR (get_value))
+#if ENABLED (JERRY_ESNEXT)
+    /* ES11:4.a. */
+    if (len + args_number > ECMA_NUMBER_MAX_SAFE_INTEGER)
     {
-      return get_value;
+      return ecma_raise_type_error (ECMA_ERR_MSG ("Unshift elements over 2**53-1 length is disallowed"));
     }
+#endif /* ENABLED (JERRY_ESNEXT) */
 
-    /* 6.b */
-    ecma_number_t new_idx = ((ecma_number_t) k) + ((ecma_number_t) args_number) - 1;
-    ecma_string_t *index_str_p = ecma_new_ecma_string_from_number (new_idx);
-    ecma_value_t operation_value;
-
-    if (ecma_is_value_found (get_value))
-    {
-      /* 6.d.i, 6.d.ii */
-      operation_value = ecma_op_object_put (obj_p, index_str_p, get_value, true);
-      ecma_free_value (get_value);
-    }
-    else
+    /* ES5.1:5.,6. ES11: 4.b, 4.c */
+    for (ecma_length_t k = len; k > 0; k--)
     {
-      /* 6.e.i */
-      operation_value = ecma_op_object_delete (obj_p, index_str_p, true);
-    }
+      /* ES5.1:6.a, 6.c, ES11:4.c.i., 4.c.iii.  */
+      ecma_value_t get_value = ecma_op_object_find_by_index (obj_p, k - 1);
 
-    ecma_deref_ecma_string (index_str_p);
+      if (ECMA_IS_VALUE_ERROR (get_value))
+      {
+        return get_value;
+      }
 
-    if (ECMA_IS_VALUE_ERROR (operation_value))
-    {
-      return operation_value;
+      /* ES5.1:6.b, ES11:4.c.ii. */
+      ecma_number_t new_idx = ((ecma_number_t) k) + ((ecma_number_t) args_number) - 1;
+      ecma_string_t *index_str_p = ecma_new_ecma_string_from_number (new_idx);
+      ecma_value_t operation_value;
+
+      if (ecma_is_value_found (get_value))
+      {
+        /* ES5.1:6.d.i, 6.d.ii, ES11:4.c.iv. */
+        operation_value = ecma_op_object_put (obj_p, index_str_p, get_value, true);
+        ecma_free_value (get_value);
+      }
+      else
+      {
+        /* ES5.1:6.e.i, ES11:4.c.v. */
+        operation_value = ecma_op_object_delete (obj_p, index_str_p, true);
+      }
+
+      ecma_deref_ecma_string (index_str_p);
+
+      if (ECMA_IS_VALUE_ERROR (operation_value))
+      {
+        return operation_value;
+      }
     }
   }
 
   for (uint32_t arg_index = 0; arg_index < args_number; arg_index++)
```

Comment:
Shouldn't this loop also be in the `should_iterate` block?

Reply:
You're right, it should because of the spec. I'll fix it immediately.
Technically it is no-op, because args_number is still 0. 
But let's follow the spec and avoid double checking of args_number.